### PR TITLE
8298873: Update IllegalRecordVersion.java for changes to TLS implementation

### PR DESCRIPTION
**Pull request by mpdonova**
- Updated ProtocolVersion.isNegotiable() to check a bounded range of version numbers.
- Removed IllegalRecordVersion.java from ProblemList.txt 

Tested with jdk_security and jdk_security3 test groups.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298873](https://bugs.openjdk.org/browse/JDK-8298873): Update IllegalRecordVersion.java for changes to TLS implementation

### Reviewers
 * [Rajan Halade](https://openjdk.org/census#rhalade) (@rhalade - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11929/head:pull/11929` \
`$ git checkout pull/11929`

Update a local copy of the PR: \
`$ git checkout pull/11929` \
`$ git pull https://git.openjdk.org/jdk pull/11929/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11929`

View PR using the GUI difftool: \
`$ git pr show -t 11929`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11929.diff">https://git.openjdk.org/jdk/pull/11929.diff</a>

</details>

### CODE REVIEW

**Review by XueleiFan — CHANGES_REQUESTED** [review]
This update will introduce version negotiation issues.    Per TLS spec, version 105.106 should be allowed and the version could be negotiated properly.  When TLS 1.4 is defined later in the future, the code update here will cause serious compatibility issues.  This has been a well-known issue in some implementations.

If you want to fix the javax/net/ssl/SSLEngine/IllegalRecordVersion.java issue, please refer to the JDK-8042449 patch details.

**Comment by mpdonova** [comment]
> This update will introduce version negotiation issues. Per TLS spec, version 105.106 should be allowed and the version could be negotiated properly. When TLS 1.4 is defined later in the future, the code update here will cause serious compatibility issues. This has been a well-known issue in some implementations.

I'm not sure what you mean here. Can you point me towards the spec that you're referring to? 

If we need to support later, currently undefined, versions then is IllegalRecordVersion a valid test?

> 
> If you want to fix the javax/net/ssl/SSLEngine/IllegalRecordVersion.java issue, please refer to the JDK-8042449 patch details.

I looked at that original patch and the code history after that. The original patch (JDK-8042449) created/updated `InputRecord.checkRecordVersion()` But then in this [commit](https://github.com/openjdk/jdk/commit/87c6761704ad5ae5f03c0cbc4de33a97ddb9317c#diff-0598abb15a2271a12c36131833ff938ab9d841b1ed69bc1b4eeabfcc460067d8) the `checkRecordVersion()` method was removed entirely and any version-checking was moved to `ProtocolVersion.isNegotiable()`.

The original `checkRecordVersion()` is:
```
if ((version.v < ProtocolVersion.MIN.v) ||
            ((version.major & 0xFF) > (ProtocolVersion.MAX.major & 0xFF))) {

            // if it's not SSLv2, we're out of here.
            if (!allowSSL20Hello ||
                    (version.v != ProtocolVersion.SSL20Hello.v)) {
                throw new SSLException("Unsupported record version " + version);
            }
        }
```

The logic is structured differently, but I'm pretty sure that it's has the same result.

**Comment by XueleiFan** [comment]
> > This update will introduce version negotiation issues. Per TLS spec, version 105.106 should be allowed and the version could be negotiated properly. When TLS 1.4 is defined later in the future, the code update here will cause serious compatibility issues. This has been a well-known issue in some implementations.
> 
> I'm not sure what you mean here. Can you point me towards the spec that you're referring to?
> 
Please refer to ["Appendix E.  Backward Compatibility"](https://www.rfc-editor.org/rfc/rfc5246#appendix-E) of RFC 5246.  Let see an example, suppose TLS 1.4 is defined. If the server is only able to accept TLS 1.3, if the client is using TLS 1.4 format, the connection cannot be established.  But TLS 1.3 should be negotiated.

BTW, this filed has been deprecated and "MUST be ignored for all purposes" since TLS 1.3 (See RFC 8446).

> If we need to support later, currently undefined, versions then is IllegalRecordVersion a valid test?
That's the good question.  It may worthy of further evaluation and the test case could be removed if it is not valid.

File: test/jdk/ProblemList.txt [hunk 1]
```diff
@@ -587,7 +587,6 @@ sun/security/pkcs11/sslecc/ClientJSSEServerJSSE.java            8161536 generic-
 sun/security/tools/keytool/ListKeychainStore.sh                 8156889 macosx-all
 
 javax/net/ssl/SSLEngine/TestAllSuites.java                      8298874 generic-all
-javax/net/ssl/SSLEngine/IllegalRecordVersion.java               8298873 generic-all
```
Comment (rhalade) [inline test/jdk/ProblemList.txt:590]: I have closed 8298873 as duplicate of this bug. Can you please update IllegalRecordVersion test to list 8299870 under `@bug`.

**Comment by mpdonova** [comment]
> > If we need to support later, currently undefined, versions then is IllegalRecordVersion a valid test?
> That's the good question. It may worthy of further evaluation and the test case could be removed if it is not valid.

I reworked `IllegalRecordVersion.java` so that it creates a ClientHello with a bad value in that version field, continues the handshake to the end, and then verifies that a version was agreed upon.

If that sounds legitimate, I can clean up the code a little and push it.

**Comment by XueleiFan** [comment]
> > > If we need to support later, currently undefined, versions then is IllegalRecordVersion a valid test?
> > > That's the good question. It may worthy of further evaluation and the test case could be removed if it is not valid.
> 
> I reworked `IllegalRecordVersion.java` so that it creates a ClientHello with a bad value in that version field, continues the handshake to the end, and then verifies that a version was agreed upon.
> 
> If that sounds legitimate, I can clean up the code a little and push it.

I'm fine for this approach.  The file/class name could be revised.

File: test/jdk/javax/net/ssl/SSLEngine/HandshakeWithInvalidRecordVersion.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    private static final String PATH_TO_STORES = "../etc";
+    private static final String KEYSTORE_FILE = "keystore";
+    private static final String TRUSTSTORE_FILE = "truststore";
+
+    private static final String KEYSTORE_PATH =
+            System.getProperty("test.src", "./") + "/" + PATH_TO_STORES +
+                    "/" + KEYSTORE_FILE;
+    private static final String TRUSTSTORE_PATH =
+            System.getProperty("test.src", "./") + "/" + PATH_TO_STORES +
+                    "/" + TRUSTSTORE_FILE;
```
Comment (XueleiFan) [inline test/jdk/javax/net/ssl/SSLEngine/HandshakeWithInvalidRecordVersion.java:51]: It would be nice to avoid to use binary files.  Would you mind to check if test/jdk/javax/net/ssl/templates/SSLContextTemplate.java could be used for the generation of SSLContext (for example test/jdk/javax/net/ssl/templates/SSLEngineTemplate.java)?

**Review by XueleiFan — COMMENTED** [review]
Would you please update the subject and description of JDK-8299870 so that it fit better with the purpose of the patch?

File: src/java.base/share/classes/sun/security/ssl/ProtocolVersion.java [hunk 3]
```diff
@@ -407,4 +407,4 @@ static ProtocolVersion selectedFrom(
 
         return selectedVersion;
     }
-}
+}
```
Comment (rhalade) [inline src/java.base/share/classes/sun/security/ssl/ProtocolVersion.java:410]: revert this change as well.

File: test/jdk/ProblemList.txt (hunk 1 above)
Comment (rhalade) [reply test/jdk/ProblemList.txt:590]: Since 8299870 is test only bug now, we don't need to list it.

**Comment by rhalade** [comment]
Thinking may be it is better to close JDK-8299870 and "Not an Issue" for archival purpose and address the test fix as part of  JDK-8298873. What do you think?

**Review by rhalade — APPROVED** [review]

**Comment by RogerRiggs** [comment]
The copyright line has the incorrect syntax. Please followup with a fix.
Missing comma, after 2023.

File: test/jdk/javax/net/ssl/SSLEngine/HandshakeWithInvalidRecordVersion.java (hunk 2 above)
Comment (rhalade) [reply test/jdk/javax/net/ssl/SSLEngine/HandshakeWithInvalidRecordVersion.java:52]: There are few other tests that need update as well. We should create backlog item to update all these tests to use the SSLEngineTemplate to generate SSLContext.
